File: Chapter3/Python/Example4.py
# Chapter 3 - Binomial Distribution
# Example 4: Simulation  B(50, 0.3) — 1000 experiments

import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import binom

# Parameters
n               = 50
p               = 0.3
num_simulations = 1000

# Simulation
np.random.seed(42)
simulated_data = binom.rvs(n, p, size=num_simulations)

# Theoretical distribution
x               = np.arange(0, n + 1)
theoretical_pmf = binom.pmf(x, n, p)

# Plot
plt.figure(figsize=(12, 6))

plt.hist(simulated_data, bins=np.arange(-0.5, n + 1.5, 1),
         density=True, alpha=0.6, color='blue',
         edgecolor='black', label='Simulation')

plt.plot(x, theoretical_pmf, 'ro-', linewidth=2,
         markersize=6, label='Theoretical Distribution')

plt.xlabel('Number of Successes', fontsize=12)
plt.ylabel('Frequency / Probability', fontsize=12)
plt.title(f'Simulation vs Theoretical Distribution: B({n}, {p})', fontsize=14)
plt.legend(fontsize=11)
plt.grid(alpha=0.3)
plt.tight_layout()
plt.show()

# Statistics comparison
print(f"Simulation Mean:       {np.mean(simulated_data):.2f}")
print(f"Theoretical Mean:      {n * p:.2f}")
print(f"Simulation Std Dev:    {np.std(simulated_data, ddof=1):.2f}")
print(f"Theoretical Std Dev:   {np.sqrt(n * p * (1 - p)):.2f}")

File: Chapter3/R/Example4.R
# Chapter 3 - Binomial Distribution
# Example 4: Simulation  B(50, 0.3) — 1000 experiments

# Parameters
n               <- 50
p               <- 0.3
num_simulations <- 1000

# Simulation
set.seed(42)
simulated_data <- rbinom(num_simulations, n, p)

# Theoretical distribution
x               <- 0:n
theoretical_pmf <- dbinom(x, n, p)

# Plot
hist(simulated_data, breaks = seq(-0.5, n + 0.5, 1),
     freq   = FALSE, col = 'lightblue', border = 'black',
     xlab   = 'Number of Successes',
     ylab   = 'Frequency / Probability',
     main   = paste0('Simulation vs Theoretical Distribution: B(', n, ', ', p, ')'))

# Overlay theoretical line
lines(x, theoretical_pmf, col = 'red', lwd = 2, type = 'b', pch = 19)

legend('topright',
       legend = c('Simulation', 'Theoretical Distribution'),
       fill   = c('lightblue', NA), border = c('black', NA),
       col    = c(NA, 'red'), lwd = c(NA, 2), pch = c(NA, 19))

# Statistics comparison
cat(sprintf("Simulation Mean:       %.2f\n", mean(simulated_data)))
cat(sprintf("Theoretical Mean:      %.2f\n", n * p))
cat(sprintf("Simulation Std Dev:    %.2f\n", sd(simulated_data)))
cat(sprintf("Theoretical Std Dev:   %.2f\n", sqrt(n * p * (1 - p))))
